dvc/project.py

```diff
@@ -165,6 +165,14 @@ def install(self):
     def to_dvc_path(self, path):
         return os.path.relpath(path, self.root_dir)
 
+    def _check_cwd_specified_as_output(self, cwd):
+        from dvc.exceptions import WorkingDirectoryAsOutputError
+
+        for stage in self.stages():
```

Comment:
Note that we call self.stages() here as well as in other checks such as _check_output_duplication. That method goes and collects all the stages all over again, which takes quite a lot of time in big projects. We need to re-use self.stages() in these checks.

Reply:
@efiop , I thought about caching the `self.stages()`in something like: https://stackoverflow.com/questions/41228551/storing-calculated-values-in-an-object/41230310#41230310

But thinking about it will affect any future library usage; something quick is just to store the value in a variable on the scope of `run` and pass it as an argument to those `_check` functions, what do you think?

Reply:
> I thought about caching the self.stages()in something like: https://stackoverflow.com/questions/41228551/storing-calculated-values-in-an-object/41230310#41230310

We've had `cached_property` used before, but in this particular case stages() method didn't require caching, since it is sometimes used to dynamically update stages.

> But thinking about it will affect any future library usage; something quick is just to store the value in a variable on the scope of run and pass it as an argument to those _check functions, what do you think?

Yes, that would be great! :slightly_smiling_face: